bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java

```diff
@@ -1080,7 +1083,19 @@ journalFormatVersionToWrite, getBufferedChannelBuilder(),
                             .registerSuccessfulEvent(MathUtils.elapsedNanos(dequeueStartTime), TimeUnit.NANOSECONDS);
                     }
 
-                    if (numEntriesToFlush == 0) {
+                    if (localQueueEntries.isEmpty()) {
+                        queue.drainTo(localQueueEntries);
+                    }
+
+                    if (!localQueueEntries.isEmpty()) {
```

Comment:
If new entries keep put into the `queue`, the `localQueueEntries` won't be empty, it won't go into line#1107 branch, and the flush won't be triggered.

Reply:
I'm not sure I got the scenario here. I think every time the loop goes back to the top on the `while (true)` and we should always check first the `localQueueEntries`  before attempting the `poll()`

Reply:
Very good points. I actually think we should do it both before the `poll()` and the `take()`. 

@zymap @hangc0276  PTAL again